### PR TITLE
Add conformance tests for trap-expr

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/35814

### CODE REVIEW

File: conformance/lang/expressions/trap_expr.balt [hunk 1]
```diff
@@ -0,0 +1,252 @@
+Test-Case: output
+Description: Test trap-expr when the expression is a nil-literal.
+Labels: trap-expr, error-type, null, nil-literal, is-expr
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:3]: +union-type

File: conformance/lang/expressions/trap_expr.balt [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    ()|error t1 = trap ();
+    io:println(t1 is ()); // @output true
+
+    error? t2 = trap null;
+    io:println(t2 is ()); // @output true
+}
+
+Test-Case: output
+Description: Test trap-expr when the expression is a boolean-literal.
+Labels: trap-expr, error-type, boolean, boolean-literal
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:15]: +union-type

Comment (SandaruJayawardana) [reply conformance/lang/expressions/trap_expr.balt:15]: Please check the other places too.

File: conformance/lang/expressions/trap_expr.balt [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+Test-Case: output
+Description: Test trap-expr when the expression is a string-template-expr.
+Labels: trap-expr, error-type, string-template-expr
+
+function init() {
+    string|error t1 = trap string `Hello World!`;
+    io:println(t1); // @output Hello World!
+    string s = string `Hello World!`;
+    string|error t2 = string `${s}`;
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:65]: `trap` is missing

File: conformance/lang/expressions/trap_expr.balt [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+function init() {
+    int[]|error t1 = trap [1, 2, 3, 4];
+    io:println(t1); // @output [1,2,3,4]
+    (int|string)[]|error t2 = trap [1, "2", 3, "4"];
+    io:println(t2); // @output [1,"2",3,"4"]
+}
+
+Test-Case: output
+Description: Test trap-expr when the expression is a list-constructor-expr.
+Labels: trap-expr, error-type, int, string, union-type, list-constructor-expr, type-cast-expr, function-call-expr
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:91]: +return-stmt, function-defn

File: conformance/lang/expressions/trap_expr.balt [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+Labels: trap-expr, error-type, xml-template-expr
+
+function init() {
+    xml|error t1 = trap xml `<Greeting>Hello!</Greeting>`;
+    io:println(t1); // @output <Greeting>Hello!</Greeting>
+}
+
+Test-Case: output
+Description: Test trap-expr when the expression is a list-constructor-expr.
+Labels: trap-expr, error-type, int, string, union-type, list-constructor-expr
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:80]: +array-type

File: conformance/lang/expressions/trap_expr.balt (hunk 4 above)
Comment (SandaruJayawardana) [reply conformance/lang/expressions/trap_expr.balt:91]: +array-type

File: conformance/lang/expressions/trap_expr.balt [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+Description: Test trap-expr when the expression is a xml-template-expr.
+Labels: trap-expr, error-type, xml-template-expr
+
+function init() {
+    xml|error t1 = trap xml `<Greeting>Hello!</Greeting>`;
+    io:println(t1); // @output <Greeting>Hello!</Greeting>
+}
+
+Test-Case: output
+Description: Test trap-expr when the expression is a list-constructor-expr.
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:79]: Since there are multiple tests for `list-constructor-expr`, shall we add a test number in here?
```suggestion
Description: Test trap-expr when the expression is a list-constructor-expr. Test 1.
```

File: conformance/lang/expressions/trap_expr.balt [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    return "2";
+}
+
+function init() {
+    int[]|error t = trap [1, 2, <int> foo(), 4];
+    io:println(t); // @output error("{ballerina}TypeCastError",message="incompatible types: 'string' cannot be cast to 'int'")
+}
+
+Test-Case: output
+Description: Test trap-expr when the expression is a list-constructor-expr.
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:103]:
```suggestion
Description: Test trap-expr when the expression is a table-constructor-expr.
```

File: conformance/lang/expressions/trap_expr.balt [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+Labels: trap-expr, error-type, int, string, union-type, table-constructor-expr, multiplicative-expr
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:104]: +record-type, record-type-readonly-field

File: conformance/lang/expressions/trap_expr.balt [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+function init() {
+    (table<record {|readonly string name; int id;|}> key(name))|error tbl1 = trap table key(name) [
+        {name: "John Doe", id: 9223372036854775807*2}
+    ];
+    io:println(tbl1); // @output error("{ballerina}NumberOverflow",message="int range overflow")
+}
+
+Test-Case: output
+Description: Test trap-expr when the expression is a object-constructor-expr.
+Labels: trap-expr, error-type, object-constructor-expr, method-call-expr, panic
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:115]: +optional-type
I think, we have to add `function-type` also?

File: conformance/lang/expressions/trap_expr.balt [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+                                            panic error("msg");
+                                        }
+                                    };
+    error? e = trap obj.bar();
+    io:println(e); // @output error("msg")
+}
+
+Test-Case: output
+Description: Test trap-expr when the expression is a new-expr.
+Labels: trap-expr, error-type, new-expr, method-call-expr, panic, error-constructor-expr
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:133]: +module-class-defn, optional-type

File: conformance/lang/expressions/trap_expr.balt [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+Test-Case: output
+Description: Test trap-expr when the expression is a type-cast-expr.
+Labels: trap-expr, error-type, int, string, function-call-expr, union-type, type-cast-expr
+
+function foo() returns int|string {
+    return "str";
+}
+
+function init() {
+    int|error t1 = trap <int>foo();
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:185]: Shall we add a test with valid type cast also?
```ballerina
int|error t2 = trap <string>foo();
```

File: conformance/lang/expressions/trap_expr.balt [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+function init() {
+    boolean|error t1 = trap !<boolean>foo();
+    io:println(t1); // @output error("{ballerina}TypeCastError",message="incompatible types: 'int' cannot be cast to 'boolean'")
+}
+
+Test-Case: output
+Description: Test trap-expr when the expression is a multiplicative-expr.
+Labels: trap-expr, error-type, int, boolean, function-call-expr, multiplicative-expr
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:204]: Shall we add a test with multiplication also?

File: conformance/lang/expressions/trap_expr.balt [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+    return 0;
+}
+
+function init() {
+    int|error t1 = trap trap 2 / foo();
+    io:println(t1); // @output error("{ballerina}DivisionByZero",message=" / by zero")
+}
+
+Test-Case: error
+Description: Check the type of trap-expr.
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:229]: Shall we add more scenarios for this? with literals etc.

File: conformance/lang/expressions/trap_expr.balt [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+Labels: trap-expr, error-type, int, boolean, function-call-expr, error-constructor-expr
+
+function funcNeverReturns() returns never {
+    panic error("msg");
+}
+
+function init() {
+    never|error e = trap funcNeverReturns();
+    io:println(e); // @output error("msg")
+}
```
Comment (SandaruJayawardana) [inline conformance/lang/expressions/trap_expr.balt:252]: We can add few more tests with `check`, `checkpanic` (including custom errors), integer subtypes, `char`

Comment (SandaruJayawardana) [reply conformance/lang/expressions/trap_expr.balt:252]: and also with `byte`

File: conformance/lang/expressions/trap_expr.balt [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+Test-Case: output
+Description: Test trap-expr when the type of expression is never.
+Labels: trap-expr, error-type, int, boolean, function-call-expr, error-constructor-expr
+
+function funcNeverReturns() returns never {
+    panic error("msg");
+}
+
+function init() {
+    never|error e = trap funcNeverReturns();
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:250]: Please add a test like
```ballerina
    error e = trap funcNeverReturns();
```
also.

File: conformance/lang/expressions/trap_expr.balt [hunk 16]
```diff
@@ -0,0 +1,252 @@
+Test-Case: output
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:1]: The structure we usually follow is to 
1. first cover the sections/sentences in the spec
2. then write different tests with different kinds of expressions

Please reorganize the tests to match that.

File: conformance/lang/expressions/trap_expr.balt [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+Description: Test trap-expr when the expression is a nil-literal.
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:2]: In the description let's use 
```suggestion
Description: Test trap expression when the expression is a nil-literal.
```

Comment (MaryamZi) [reply conformance/lang/expressions/trap_expr.balt:2]: Please fix wherever applicable for all the productions that have been used.

File: conformance/lang/expressions/trap_expr.balt (hunk 1 above)
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:3]: Labels have to in alphabetical order.

Comment (MaryamZi) [reply conformance/lang/expressions/trap_expr.balt:3]: optional-type, nil-type.

File: conformance/lang/expressions/trap_expr.balt [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+Description: Check the type of trap-expr.
+Labels: trap-expr, error-type, int, function-call-expr, multiplicative-expr
+
+function foo() returns int {
+    return 0;
+}
+
+function errorFunction() {
+    int _ = trap 2 / foo(); // @error incompatible types: expected 'int', found '(int|error)'
+    error e = trap 2 / foo(); // @error incompatible types: expected 'error', found 'int'
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:238]: Isn't the test overly complicated? We don't need a function call. Even a variable reference would suffice.

Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:238]: Is the error correct?

File: conformance/lang/expressions/trap_expr.balt [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+
+function foo() returns int {
+    return 0;
+}
+
+function init() {
+    int|error t1 = trap trap 2 / foo();
+    io:println(t1); // @output error("{ballerina}DivisionByZero",message=" / by zero")
+}
+
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:227]: Do you have tests for when the expression doesn't panic?

File: conformance/lang/expressions/trap_expr.balt (hunk 1 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 2 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Changed

File: conformance/lang/expressions/trap_expr.balt (hunk 3 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 4 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 5 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 10 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 11 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt:249]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 12 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 13 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 14 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt:333]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 15 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 16 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt:1]: Changed

File: conformance/lang/expressions/trap_expr.balt (hunk 17 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Changed

File: conformance/lang/expressions/trap_expr.balt (hunk 1 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Changed

File: conformance/lang/expressions/trap_expr.balt (hunk 18 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Changed

Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Created an issue and moved this test to failing tests

File: conformance/lang/expressions/trap_expr.balt (hunk 19 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt:289]: Added few tests at the bottom

File: conformance/lang/expressions/trap_expr.balt [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+Labels: error-type, int, local-init-var-decl-stmt, multiplicative-expr, trap-expr, variable-reference-expr
+
+function errorFunction() {
+    int zero = 0;
+    error e = trap 2 / zero; // @error incompatible types: expected 'error', found 'int'
+}
+
+Test-Case: output
+Description: Test trap expression when the type of expression is never.
+Labels: boolean, error-constructor-expr, error-type, function-call-expr, function-defn, int, return-stmt, trap-expr, union-type
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a byte-array-literal.
+Labels: byte, byte-array-literal, error-type, trap-expr, union-type
+
+function init() {
+    byte[]|error t1 = trap base16 `aeeecdefabcd12345567888822`;
+    io:println(t1); // @output [174,238,205,239,171,205,18,52,85,103,136,136,34]
+    t1 = base64 `aGVsbG8gYmFsbGVyaW5hICEhIQ==`;
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Changed

File: conformance/lang/expressions/trap_expr.balt [hunk 22]
```diff
@@ -0,0 +1,333 @@
...
+
+function init() {
+    int zero = 0;
+    int|error t1 = trap trap 2 / zero;
+    io:println(t1); // @output error("{ballerina}DivisionByZero",message=" / by zero")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a checking expression.
+Labels: boolean, error-type, int, local-init-var-decl-stmt, multiplicative-expr, trap-expr, union-type, variable-reference-expr
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt [hunk 23]
```diff
@@ -0,0 +1,333 @@
...
+Labels: boolean, error-type, int:Signed32, int:Unsigned32, local-init-var-decl-stmt, trap-expr, union-type, variable-reference-expr
+
+function init() {
+    int:Signed32 i1 = -12;
+    int:Signed32|error e1 = trap i1;
+    io:println(e1); // @output -12
+    int:Unsigned32 i2 = 1234;
+    int:Unsigned32|error e2 = trap i2;
+    io:println(e2); // @output 1234
+}
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt:333]: Added

File: conformance/lang/expressions/trap_expr.balt [hunk 24]
```diff
@@ -0,0 +1,351 @@
+Test-Case: output
+Description: Test evaluation of expression resulting in value v,
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:2]: Do we need this? It's not clear what `v` is. I would sau

```suggestion
Description: Test the result of the trap expression being `e`, when evaluation of the expression completes abruptly with panic with associated value `e`.
```

Comment (MaryamZi) [reply conformance/lang/expressions/trap_expr.balt:2]: Please recheck all descriptions. We need to make the concise, but self-explanatory and clear.

File: conformance/lang/expressions/trap_expr.balt [hunk 25]
```diff
@@ -0,0 +1,351 @@
...
+
+function init() {
+    int|error t1 = trap foo();
+    io:println(t1); // @output error("{ballerina/lang.array}IndexOutOfRange",message="array index out of range: index: 3, size: 0")
+    error? t2 = trap bar();
+    io:println(t2); // @output error("msg")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a type-cast-expr.
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:241]:
```suggestion
Description: Test trap expression when the expression is a type cast expression.
```

Please update wherever applicable.

File: conformance/lang/expressions/trap_expr.balt (hunk 19 above)
Comment (MaryamZi) [reply conformance/lang/expressions/trap_expr.balt:311]: Is it the test immediately after this? Can we improve the descriptions to specifically mention that they don't panic?

File: conformance/lang/expressions/trap_expr.balt [hunk 26]
```diff
@@ -0,0 +1,351 @@
...
+
+Test-Case: error
+Description: If type of expr is T, then type of trap expr is T|error.
+Labels: error-type, int, local-init-var-decl-stmt, multiplicative-expr, nil-literal, optional-type, string, trap-expr, union-type, variable-reference-expr
+
+function errorFunction() {
+    int zero = 0;
+    int? nilVal = ();
+    int _ = trap 2 / zero; // @error incompatible types: expected 'int', found '(int|error)'
+    string|error? x = trap nilVal; // @error incompatible types: expected '(string|error)?', found 'int?'
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:34]: Shouldn't this be
```suggestion
    string|error? x = trap nilVal; // @error incompatible types: expected '(string|error)?', found '(int|error)?'
```

File: conformance/lang/expressions/trap_expr.balt [hunk 27]
```diff
@@ -0,0 +1,351 @@
...
+    int _ = trap 2; // @error incompatible types: expected 'int', found '(int|error)'
+    string _ = trap "str"; // @error incompatible types: expected 'string', found '(string|error)'
+    float _ = trap 2.3f; // @error incompatible types: expected 'float', found '(float|error)'
+}
+
+Test-Case: error
+Description: If type of expr is T, then type of trap expr is T|error.
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:41]:
```suggestion
Description: If type of `expr` is `T`, then type of `trap expr` is `T|error`.
```

Let's fix wherever applicable.

File: conformance/lang/expressions/trap_expr.balt [hunk 28]
```diff
@@ -0,0 +1,351 @@
...
+Fail-Issue: ballerina-platform/ballerina-lang#36472
+Labels: error-type, int, local-init-var-decl-stmt, multiplicative-expr, trap-expr, variable-reference-expr
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:43]: `local-init-var-decl-stmt` is part of the minimal subset. Why are we specifically mentioning it here?

Comment (MaryamZi) [reply conformance/lang/expressions/trap_expr.balt:43]: So is `variable-reference-expr`.

File: conformance/lang/expressions/trap_expr.balt [hunk 29]
```diff
@@ -0,0 +1,351 @@
...
+
+function errorFunction() {
+    int zero = 0;
+    error e = trap 2 / zero; // @error incompatible types: expected 'error', found 'int'
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:47]: Let's add a test like `error e2 = trap zero;` also.

File: conformance/lang/expressions/trap_expr.balt [hunk 30]
```diff
@@ -0,0 +1,351 @@
...
+
+function funcNeverReturns() returns never {
+    panic error("msg");
+}
+
+function init() {
+    never|error e1 = trap funcNeverReturns();
+    io:println(e1); // @output error("msg")
+    error e2 = trap funcNeverReturns();
+    io:println(e2); // @output error("msg")
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:62]: Let's have `funcNeverReturns` accept a string that is incorporated in the error message. That way we get two error values as `e1` and `e2`.

File: conformance/lang/expressions/trap_expr.balt [hunk 31]
```diff
@@ -0,0 +1,351 @@
...
+Description: Test trap expression when the expression is a boolean-literal and it does not panic.
+Labels: boolean, boolean-literal, error-type, trap-expr, union-type
+
+function init() {
+    boolean|error t = trap false;
+    io:println(t); // @output false
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a numeric-literal and it does not panic.
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:87]: When does a numeric literal panic?

Comment (MaryamZi) [reply conformance/lang/expressions/trap_expr.balt:87]: Same for string.

File: conformance/lang/expressions/trap_expr.balt [hunk 32]
```diff
@@ -0,0 +1,351 @@
...
+
+function init() {
+    int[]|error t1 = trap [1, 2, 3, 4];
+    io:println(t1); // @output [1,2,3,4]
+    (int|string)[]|error t2 = trap [1, "2", 3, "4"];
+    io:println(t2); // @output [1,"2",3,"4"]
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a list constructor expression. Test 2.
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:154]: Isn't the difference between this and the previous the fact that one doesn't complete normally? Why aren't we saying that instead of Test 1 and Test 2?

File: conformance/lang/expressions/trap_expr.balt [hunk 33]
```diff
@@ -0,0 +1,351 @@
...
+function init() {
+    int zero = 0;
+    int|error e1 = trap (check 2 / zero);
+    io:println(e1); // @output error("{ballerina}DivisionByZero",message=" / by zero")
+    int|error e2 = trap (checkpanic 2 / zero);
+    io:println(e2); // @output error("{ballerina}DivisionByZero",message=" / by zero")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a char and it does not panic.
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:303]:
```suggestion
Description: Test trap expression when the expression is a `string:Char` and it does not panic.
```

or
```suggestion
Description: Test trap expression when the expression is a character and it does not panic.
```

File: conformance/lang/expressions/trap_expr.balt [hunk 34]
```diff
@@ -0,0 +1,351 @@
...
+Labels: boolean, error-type, local-init-var-decl-stmt, string:Char, trap-expr, union-type, variable-reference-expr
+
+function init() {
+    string:Char c = "c";
+    string:Char|error e1 = trap c;
+    io:println(e1); // @output c
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a byte value and it does not panic.
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:313]: Can we add some tests with function calls here instead of literals and variable reference expressions that we know will never panic?

Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt:313]: @MaryamZi, do you mean here to add function calls that causes panic or add functions that return literal?

File: conformance/lang/expressions/trap_expr.balt (hunk 28 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Removed them

File: conformance/lang/expressions/trap_expr.balt (hunk 29 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Added

File: conformance/lang/expressions/trap_expr.balt (hunk 30 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Changed

File: conformance/lang/expressions/trap_expr.balt (hunk 34 above)
Comment (MaryamZi) [reply conformance/lang/expressions/trap_expr.balt:305]: Like a function call, which at compile-time we don't know if it will panic.

```ballerina
function init() {
    int|error a = trap fn("1");
    io:println(a);
}

function fn(string str) returns int {
    return checkpanic int:fromString(str);
}

```

File: conformance/lang/expressions/trap_expr.balt (hunk 31 above)
Comment (MaryamZi) [reply conformance/lang/expressions/trap_expr.balt:88]: Thoughts on this? 

Let's also remove the productions from the descriptions everywhere.

numeric-literal -> numeric literal

Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Fixed in 5303f62061d4c9f05d3292185a6988ae3261a1a4

File: conformance/lang/expressions/trap_expr.balt (hunk 34 above)
Comment (KavinduZoysa) [reply conformance/lang/expressions/trap_expr.balt]: Changed

File: conformance/lang/expressions/trap_expr.balt [hunk 35]
```diff
@@ -0,0 +1,369 @@
...
+function init() {
+    byte b = 0;
+    byte|error e = trap foo(true);
+    io:println(e); // @output 0
+    e = trap foo(false);
+    io:println(e); // @output error("Don't return byte")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a function call which returns subtype of int or panic.
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:335]:
```suggestion
Description: Test trap expression when the expression is a function call which returns a subtype of int or panics.
```

File: conformance/lang/expressions/trap_expr.balt [hunk 36]
```diff
@@ -0,0 +1,369 @@
...
+
+function init() {
+    int:Signed32|error e1 = trap foo(true);
+    io:println(e1); // @output -12
+    e1 = trap foo(false);
+    io:println(e1); // @output error("Don't return signed int")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a query expression
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:354]:
```suggestion
Description: Test trap expression when the expression is a query expression.
```

File: conformance/lang/expressions/trap_expr.balt (hunk 35 above)
Comment (MaryamZi) [reply conformance/lang/expressions/trap_expr.balt:335]: Please fix wherever applicable.

File: conformance/lang/expressions/trap_expr.balt [hunk 37]
```diff
@@ -0,0 +1,369 @@
...
+    io:println(t1); // @output error("{ballerina}DivisionByZero",message=" / by zero")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a checking expression.
+Labels: boolean, check, checkpanic, error-type, int, multiplicative-expr, trap-expr, union-type
+
+function init() {
+    int zero = 0;
+    int|error e1 = trap (check 2 / zero);
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:288]: I think this is a bug. `trap` captures a panic. `check` should return the error from the function. Can you recheck in the spec and create a spec issue if required or a lang issue?

File: conformance/lang/expressions/trap_expr.balt [hunk 38]
```diff
@@ -0,0 +1,369 @@
...
+    io:println(trap i[3]); // @output error("{ballerina/lang.array}IndexOutOfRange",message="array index out of range: index: 3, size: 0")
+}
+
+Test-Case: output
+Description: Test trap expression when the expression is a function call expression.
+Labels: array-type, error-constructor-expr, error-type, function-call-expr, function-defn, int, optional-type, return-stmt, trap-expr, union-type
+
+function foo() returns int {
+    int[] i = [];
+    return i[3];
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:218]: Missing member access related labels in `Labels`.

File: conformance/lang/expressions/trap_expr.balt [hunk 39]
```diff
@@ -0,0 +1,369 @@
...
+
+class Obj {
+    function bar() {
+        panic error("msg");
+    }
+}
+
+function init() {
+    Obj obj = new();
+    error? t = trap obj.bar();
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:199]: This doesn't match the description.

> Test trap expression when the expression is a new expression.

File: conformance/lang/expressions/trap_expr.balt [hunk 40]
```diff
@@ -0,0 +1,369 @@
...
+    object { function bar(); } obj = object {
+                                        error? i = ();
+                                        function bar() {
+                                            self.i = self.foo();
+                                        }
+                                        function foo() {
+                                            panic error("msg");
+                                        }
+                                    };
+    error? e = trap obj.bar();
```
Comment (MaryamZi) [inline conformance/lang/expressions/trap_expr.balt:183]: Also doesn't match the description.
